bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/bookies/ReplicasMigrationCommand.java

```diff
@@ -0,0 +1,190 @@
+        for (String bookieId : bookieIds) {
+            bookieIdsSet.add(BookieId.parse(bookieId));
+        }
+        return bookieIdsSet;
+    }
+
+    public Set<Long> parseLedgerIds(String ledgerIdsOfMigratedReplica) {
+        Set<Long> ledgerIdSet = new HashSet<>();
+        String[] ledgerIds = ledgerIdsOfMigratedReplica.split(seperator);
+        for (String ledgerId : ledgerIds) {
+            ledgerIdSet.add(Long.parseLong(ledgerId));
+        }
+        return ledgerIdSet;
+    }
+
+    private boolean switchToReadonly(ServerConfiguration conf, ReplicasMigrationFlags flags) throws IOException {
+        if (flags.readOnly && !conf.isHttpServerEnabled()) {
+            throw new RuntimeException("Please enable http service first, config httpServerEnabled is false!");
+        }
+
+        Set<BookieId> bookieIdSet = parseBookieIds(flags.bookieIdsOfMigratedReplica);
+        for (BookieId bookieId : bookieIdSet) {
+            String bookieAddress = bookieId.getId().split(":")[0];
+            String httpAddress = bookieAddress + ":" + flags.port;
+            String url = String.format("http://%s/%s", httpAddress, HttpRouter.BOOKIE_STATE_READONLY);
+            String readOnly = flags.readOnly ? "true" : "false";
+            CloseableHttpResponse response = HttpClientUtils.doPutRequest(url, readOnly, "UTF-8");
+            int responseCode = response.getStatusLine().getStatusCode();
+            if (responseCode != HttpServer.StatusCode.OK.getValue()) {
+                LOG.error(String.format("Set readonly to %s failed! bookieId:%s, responseCode:%s ",
+                        readOnly, bookieId, responseCode));
+                return false;
+            }
+            LOG.info(String.format("Set readonly to %s success! bookieId:%s, responseCode:%s",
+                    readOnly, bookieId, responseCode));
+        }
+        return true;
+    }
+
+    private boolean replicasMigration(ServerConfiguration conf, ReplicasMigrationFlags flags)
+            throws BKException, InterruptedException, IOException {
+        ClientConfiguration adminConf = new ClientConfiguration(conf);
+        BookKeeperAdmin admin = new BookKeeperAdmin(adminConf);
+        try {
+            Map<Long, Set<BookieId>> toMigratedLedgerAndBookieMap = new HashMap<>();
+            Set<BookieId> bookieIds = parseBookieIds(flags.bookieIdsOfMigratedReplica);
+            Set<Long> ledgerIdsToMigrate = null;
+            if (!flags.ledgerIdsOfMigratedReplica.equals(ALL)) {
+                ledgerIdsToMigrate = parseLedgerIds(flags.ledgerIdsOfMigratedReplica);
+            }
+            Map<BookieId, Set<Long>> bookieLedgerMaps = admin.listLedgers(bookieIds);
+            for (Map.Entry<BookieId, Set<Long>> bookieIdSetEntry : bookieLedgerMaps.entrySet()) {
+                Set<Long> ledgersOnBookieId = bookieIdSetEntry.getValue();
+                if (!flags.ledgerIdsOfMigratedReplica.equals(ALL)) {
```

Comment:
We'd better check 'ALL' first to reduce the call of `admin.listLedgers()`

Reply:
admin.listLedgers(bookieIds) needs to be called whether it is ALL or not. It is currently used to build the mapping structure of ledger and bookies: toMigratedLedgerAndBookieMap.

Reply:
@hangc0276

admin.listLedgers(bookieIdsSet) is for building toMigratedLedgerAndBookieMap.

toMigratedLedgerAndBookieMap records the mapping of ledger and bookie for each replica to be migrated.

Reply:
Check 'ALL' first, does not reduce the call of admin.listLedgers()